File: src/backend/vendas/src/model/nota_fiscal_tipo_model.py
from src import db
from src.model.nota_fiscal_modelo_model import NotaFiscalModeloModel


class NotaFiscalTipoModel(db.Model):
    __tablename__ = 'nota_fiscal_tipo'

    id = db.Column(db.Integer, primary_key=True)
    nome = db.Column(db.String(50))
    descricao = db.Column(db.Text)
    serie = db.Column(db.String(3))
    serie_scan = db.Column(db.String(3))
    ultimo_numero = db.Column(db.Integer)
    id_nota_fiscal_modelo = db.Column(db.Integer, db.ForeignKey('nota_fiscal_modelo.id'))

    nota_fiscal_modelo_model = db.relationship('NotaFiscalModeloModel', foreign_keys=[id_nota_fiscal_modelo])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_nota_fiscal_modelo = data.get('idNotaFiscalModelo')
        self.nome = data.get('nome')
        self.descricao = data.get('descricao')
        self.serie = data.get('serie')
        self.serie_scan = data.get('serieScan')
        self.ultimo_numero = data.get('ultimoNumero')

    def serialize(self):
        return {
            'id': self.id,
            'idNotaFiscalModelo': self.id_nota_fiscal_modelo,
            'nome': self.nome,
            'descricao': self.descricao,
            'serie': self.serie,
            'serieScan': self.serie_scan,
            'ultimoNumero': self.ultimo_numero,
            'notaFiscalModeloModel': self.nota_fiscal_modelo_model.serialize() if self.nota_fiscal_modelo_model else None,
        }

File: frontend/vendas/lib/app/data/model/nota_fiscal_tipo_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:vendas/app/data/model/model_imports.dart';

class NotaFiscalTipoModel {
	int? id;
	int? idNotaFiscalModelo;
	String? nome;
	String? descricao;
	String? serie;
	String? serieScan;
	int? ultimoNumero;
	NotaFiscalModeloModel? notaFiscalModeloModel;

	NotaFiscalTipoModel({
		this.id,
		this.idNotaFiscalModelo,
		this.nome,
		this.descricao,
		this.serie,
		this.serieScan,
		this.ultimoNumero,
		this.notaFiscalModeloModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'nome',
		'descricao',
		'serie',
		'serie_scan',
		'ultimo_numero',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Nome',
		'Descricao',
		'Serie',
		'Serie Scan',
		'Ultimo Numero',
	];

	NotaFiscalTipoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNotaFiscalModelo = jsonData['idNotaFiscalModelo'];
		nome = jsonData['nome'];
		descricao = jsonData['descricao'];
		serie = jsonData['serie'];
		serieScan = jsonData['serieScan'];
		ultimoNumero = jsonData['ultimoNumero'];
		notaFiscalModeloModel = jsonData['notaFiscalModeloModel'] == null ? NotaFiscalModeloModel() : NotaFiscalModeloModel.fromJson(jsonData['notaFiscalModeloModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNotaFiscalModelo'] = idNotaFiscalModelo != 0 ? idNotaFiscalModelo : null;
		jsonData['nome'] = nome;
		jsonData['descricao'] = descricao;
		jsonData['serie'] = serie;
		jsonData['serieScan'] = serieScan;
		jsonData['ultimoNumero'] = ultimoNumero;
		jsonData['notaFiscalModeloModel'] = notaFiscalModeloModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNotaFiscalModelo = plutoRow.cells['idNotaFiscalModelo']?.value;
		nome = plutoRow.cells['nome']?.value;
		descricao = plutoRow.cells['descricao']?.value;
		serie = plutoRow.cells['serie']?.value;
		serieScan = plutoRow.cells['serieScan']?.value;
		ultimoNumero = plutoRow.cells['ultimoNumero']?.value;
		notaFiscalModeloModel = NotaFiscalModeloModel();
		notaFiscalModeloModel?.modelo = plutoRow.cells['notaFiscalModeloModel']?.value;
	}	

	NotaFiscalTipoModel clone() {
		return NotaFiscalTipoModel(
			id: id,
			idNotaFiscalModelo: idNotaFiscalModelo,
			nome: nome,
			descricao: descricao,
			serie: serie,
			serieScan: serieScan,
			ultimoNumero: ultimoNumero,
		);			
	}

	
}